primitives/transaction-storage-proof/src/lib.rs

```diff
@@ -0,0 +1,243 @@
+	/// Get the proof.
+	fn storage_proof(&self) -> Result<Option<TransactionStorageProof>, Error>;
+}
+
+impl TransactionStorageProofInherentData for InherentData {
+	fn storage_proof(&self) -> Result<Option<TransactionStorageProof>, Error> {
+		Ok(self.get_data(&INHERENT_IDENTIFIER)?)
+	}
+}
+
+/// Provider for inherent data.
+#[cfg(feature = "std")]
+pub struct InherentDataProvider<F> {
+	inner: F,
+}
+
+#[cfg(feature = "std")]
+impl<F> InherentDataProvider<F> {
+	pub fn new(proof_oracle: F) -> Self {
+		InherentDataProvider { inner: proof_oracle }
+	}
+}
+
+#[cfg(feature = "std")]
+impl<F> sp_inherents::ProvideInherentData for InherentDataProvider<F>
+where F: Fn() -> Result<Option<TransactionStorageProof>, Error>
+{
+	fn inherent_identifier(&self) -> &'static InherentIdentifier {
+		&INHERENT_IDENTIFIER
+	}
+
+	fn provide_inherent_data(&self, inherent_data: &mut InherentData) -> Result<(), Error> {
+		let proof = (self.inner)()?;
+		if let Some(proof) = proof {
+			inherent_data.put_data(INHERENT_IDENTIFIER, &proof)
+		} else {
+			Ok(())
+		}
+	}
+
+	fn error_to_string(&self, _error: &[u8]) -> Option<String> {
+		Some(format!("no further information"))
+	}
+}
+
+/// An utility function to extract chunk index from the source of randomness.
+pub fn random_chunk(random_hash: &[u8], total_chunks: u32) -> u32 {
+	let mut buf = [0u8; 8];
+	buf.copy_from_slice(&random_hash[0..8]);
+	let random_u64 = u64::from_be_bytes(buf);
+	(random_u64 % total_chunks as u64) as u32
+}
+
+/// An utility function to enocde transaction index as trie key.
+pub fn encode_index(input: u32) -> Vec<u8> {
+	codec::Encode::encode(&codec::Compact(input))
+}
+
+#[cfg(feature = "std")]
+pub mod registration {
+	use sp_consensus::SelectChain;
+	use sp_inherents::{InherentDataProviders};
+	use log::warn;
+	use sc_client_api::{HeaderBackend, BlockBackend};
+	use sp_runtime::{traits::{Block as BlockT, Header, Saturating, Zero}, generic::BlockId};
+	use std::sync::Arc;
+	use sp_trie::TrieMut;
+	use super::*;
+
+	type Hasher = sp_core::Blake2Hasher;
+	type TrieLayout = sp_trie::Layout::<Hasher>;
+
+	/// Register uncles inherent data provider, if not registered already.
+	pub fn register_storage_proof_inherent_data_provider<B, C, SC>(
+		client: Arc<C>,
+		select_chain: SC,
+		inherent_data_providers: &InherentDataProviders,
+	) -> Result<(), sp_consensus::Error> where
+	B: BlockT,
+	C: BlockBackend<B> + HeaderBackend<B> + Send + Sync + 'static,
+	SC: SelectChain<B> + 'static,
+	{
+		if !inherent_data_providers.has_provider(&INHERENT_IDENTIFIER) {
+			inherent_data_providers
+				.register_provider(InherentDataProvider::new(move || {
+					{
+						let chain_head = match select_chain.best_chain() {
+							Ok(x) => x,
+							Err(e) => {
+								warn!(target: "storage-proof", "Unable to get chain head: {:?}", e);
+								return Ok(None);
+							}
+						};
+
+						let number = chain_head.number().saturating_sub(DEFAULT_STORAGE_PERIOD.into());
```

Comment:
Is it fine to use directly 'DEFAULT_STORAGE_PERIOD' or should we fetch it from the runtime module (at this time nothing change it but it could?)

Reply:
As I've mentioned in the PR description, that's undecided for now. We'll probably remove storage period as config option for now.